draft-ietf-masque-connect-udp-listen.md

```diff
@@ -154,43 +171,177 @@ UDP Payload:
+rejected, the client and proxy MUST use an uncompressed context ID
+(See {{uncompressed}}) to exhange UDP payloads for the given target.
+
+### Uncompressed datagrams {#uncompressed}
+
+If the client wishes to allocate a Context ID for uncompressed packets,
+it MUST first exchange the COMPRESSION_ASSIGN capsule
+(see {{capsuleassignformat}}) with the proxy with an unused Context ID
+defined in {{contextid}} with the IP Length set to zero.
+
+
+### Compression Mapping {#mappings}
+
+When an endpoint receives a COMPRESSION_ASSIGN capsule with a non-zero IP
+length, it MUST decide whether to accept or reject the compression mapping:
+
+if it accepts the mapping, first the receiver MUST save the mapping from
+context ID to address and port. Second, the receiver MUST echo an identical COMPRESSION_ASSIGN capsule back to its peer.
+
+if it rejects the mapping, the receiver MUST respond by sending a
+COMPRESSION_CLOSE capsule with the context ID set to the one from the received COMPRESSION_ASSIGN capsule
+
+The client or proxy MAY choose to close any context contexts that it registered
+or was registered with it respectively using COMPRESSION_CLOSE
+(For example when a mapping is unused for a long time). Another potential use
+is {{restrictingips}}.
+
+
+## Restricting IPs {#restrictingips}
+
+If an uncompressed Context ID was set (via {{uncompressed}}), the client MAY at
+any point request the proxy reject all traffic from uncompressed targets by
+using COMPRESSION_CLOSE (see {{compressionclose}}) on said Context ID.
+targets effectively acting as a firewall against unwanted or unknown IPs.
+
+
+## Capsules {#capsules}
+
+The Listener capsule types are defined as follows:
+
+### The COMPRESSION_ASSIGN capsule {#compressionassign}
+
+The Compression Assign capsule has two purposes. Either to request the
+assignment of a Context ID (see {{contextid}}) to a corresponding target IP:Port.
+Or to accept a COMPRESSION_ASSIGN request from the other party.
+
+~~~ ascii-art
+Capsule {
+  Type COMPRESSION_ASSIGN (0x1C0FE323),
+  Length (i),
+  Target Information,
+}
+~~~
+{: #capsuleassignformat title="Compression Assign Capsule Format"}
+
+~~~
+Target Information {
+  Context ID (i),
+  IP Version (8),
+  [IP Address (32..128)],
+  [UDP Port (16)],
+}
+~~~
+{: #targetmappingformat title="Target Information Format"}
+
+The IP Length, Address and Port fields in {{targetmappingformat}} are the
+same as those defined in {{format}}. However, the IP version can be set
+to 0 when allocating an uncompressed Context ID, as defined in {{contextid}}.
+
+### The COMPRESSION_CLOSE capsule {#compressionclose}
+
+The Compression Close capsule serves the following purposes. As a response to
+reject a COMPRESSION_ASSIGN request and to close or clean up any existing
+compression mappings. Once a COMPRESSION_CLOSE has been exchanged between
+endpoints, they MUST NOT use that Context ID until it is re-assigned via a
```

Comment:
So as part of the debate around RFC 9298, we ended up deciding that re-registrating context IDs wasn't possible. (I didn't agree with that at the time). So to make life easier, I'd recommend removing "until it is re-assigned via a COMPRESSION_ASSIGN exchange." here. We could fight it later, but let's avoid that debate for now.

Reply:
Okay sounds good